src/isaac_system.cpp

```diff
@@ -0,0 +1,249 @@
+
+  // Initial command values
+  for (auto i = 0u; i < info_.joints.size(); i++)
+  {
+    const auto& component = info_.joints[i];
+    for (const auto& interface : component.state_interfaces)
+    {
+      auto it = std::find(standard_interfaces_.begin(), standard_interfaces_.end(), interface.name);
+      // If interface name is found in the interfaces list
+      if (it != standard_interfaces_.end())
+      {
+        auto index = static_cast<std::size_t>(std::distance(standard_interfaces_.begin(), it));
+        // Check the initial_value param is used
+        if (!interface.initial_value.empty())
+        {
+          joint_commands_[index][i] = std::stod(interface.initial_value);
+        }
+      }
+    }
+  }
+
+  // Search for mimic joints
+  for (auto i = 0u; i < info_.joints.size(); ++i)
+  {
+    const auto& joint = info_.joints.at(i);
+    if (joint.parameters.find("mimic") != joint.parameters.cend())
+    {
+      const auto mimicked_joint_it = std::find_if(
+          info_.joints.begin(), info_.joints.end(),
+          [&mimicked_joint = joint.parameters.at("mimic")](const hardware_interface::ComponentInfo& joint_info) {
+            return joint_info.name == mimicked_joint;
+          });
+      if (mimicked_joint_it == info_.joints.cend())
+      {
+        throw std::runtime_error(std::string("Mimicked joint '") + joint.parameters.at("mimic") + "' not found");
+      }
+      MimicJoint mimic_joint;
+      mimic_joint.joint_index = i;
+      mimic_joint.mimicked_joint_index =
+          static_cast<std::size_t>(std::distance(info_.joints.begin(), mimicked_joint_it));
+      auto param_it = joint.parameters.find("multiplier");
+      if (param_it != joint.parameters.end())
+      {
+        mimic_joint.multiplier = std::stod(joint.parameters.at("multiplier"));
+      }
+      mimic_joints_.push_back(mimic_joint);
+    }
+  }
+
+  const auto get_hardware_parameter = [this](const std::string& parameter_name, const std::string& default_value) {
+    if (auto it = info_.hardware_parameters.find(parameter_name); it != info_.hardware_parameters.end())
+    {
+      return it->second;
+    }
+    return default_value;
+  };
+
+  // Add random ID to prevent warnings about multiple publishers within the same node
+  rclcpp::NodeOptions options;
+  options.arguments({ "--ros-args", "-r", "__node:=isaac_ros2_control_" + info_.name });
+  auto node = rclcpp::Node::make_shared("_", options);
+
+  node_ = rclcpp::Node::make_shared("_", options);
+  isaac_joint_commands_publisher_ = node_->create_publisher<sensor_msgs::msg::JointState>(
+      get_hardware_parameter("joint_commands_topic", "/joint_command"), rclcpp::QoS(1));
+  isaac_joint_states_subscriber_ = node_->create_subscription<sensor_msgs::msg::JointState>(
+      get_hardware_parameter("joint_states_topic", "/isaac_joint_states"), rclcpp::SensorDataQoS(),
+      [this](const sensor_msgs::msg::JointState::SharedPtr joint_state) { latest_joint_state_ = *joint_state; });
```

Comment:
Don't we need a mutex around `latest_joint_state_` so we don't read and write at the same time?

Reply:
No, there's no need, there's no executor running in the background calling the callbacks, I'm explicitly spinning when needed just before using it